File: compiler_test/resources/parse/py/TAC/TAC.py
def localName(name, valueType="int"):
    if valueType.lower() == "int":
        return "li" + name
    elif valueType.lower() == "char":
        return "lc" + name


def globalName(name, valueType="int"):
    if valueType.lower() == "int":
        return "gi" + name
    elif valueType.lower() == "char":
        return "gc" + name


class TAC:
    def __init__(self, op="TACUNDEFINED", ad1="", ad2="", ad3=""):
        self.op = op
        self.ad1 = ad1
        self.ad2 = ad2
        self.ad3 = ad3

    def toDict(self):
        return {
            "op": str(self.op),
            "ad1": str(self.ad1),
            "ad2": str(self.ad2),
            "ad3": str(self.ad3),
        }


def add(ad1, ad2, ad3) -> TAC:
    return TAC("TACADD", ad1, ad2, ad3)


def sub(ad1, ad2, ad3) -> TAC:
    return TAC("TACSUB", ad1, ad2, ad3)


def mul(ad1, ad2, ad3) -> TAC:
    return TAC("TACMUL", ad1, ad2, ad3)


def div(ad1, ad2, ad3) -> TAC:
    return TAC("TACDIV", ad1, ad2, ad3)


def mov(ad1, ad3) -> TAC:
    return TAC("TACMOV", ad1, "", ad3)


def getArr(ad1, ad2, ad3) -> TAC:
    return TAC("TACGETARR", ad1, ad2, ad3)


def setArr(ad1, ad2, ad3) -> TAC:
    return TAC("TACSETARR", ad1, ad2, ad3)


def jle(ad1, ad2, ad3) -> TAC:
    return TAC("TACJLE", ad1, ad2, ad3)


def jge(ad1, ad2, ad3) -> TAC:
    return TAC("TACJGE", ad1, ad2, ad3)


def jl(ad1, ad2, ad3) -> TAC:
    return TAC("TACJL", ad1, ad2, ad3)


def jg(ad1, ad2, ad3) -> TAC:
    return TAC("TACJG", ad1, ad2, ad3)


def jne(ad1, ad2, ad3) -> TAC:
    return TAC("TACJNE", ad1, ad2, ad3)


def jeq(ad1, ad2, ad3) -> TAC:
    return TAC("TACJEQ", ad1, ad2, ad3)


def call(ad3) -> TAC:
    return TAC("TACCALL", "", "", ad3)


def ret() -> TAC:
    return TAC("TACRET", "", "", "")


def getRet(ad3) -> TAC:
    return TAC("TACGETRET", "", "", ad3)


def setRet(ad3) -> TAC:
    return TAC("TACSETRET", "", "", ad3)


def setArg(ad1, ad3) -> TAC:
    return TAC("TACSETARG", ad1, "", ad3)


def getArg(ad1, ad3) -> TAC:
    return TAC("TACGETARG", ad1, "", ad3)


def j(ad3) -> TAC:
    return TAC("TACJ", "", "", ad3)


def readint(ad3) -> TAC:
    return TAC("TACREADINT", "", "", ad3)


def readchar(ad3) -> TAC:
    return TAC("TACREADCHAR", "", "", ad3)


def writeint(ad3) -> TAC:
    return TAC("TACWRITEINT", "", "", ad3)


def writechar(ad3) -> TAC:
    return TAC("TACWRITECHAR", "", "", ad3)


def writestring(ad3) -> TAC:
    return TAC("TACWRITESTRING", "", "", ad3)


def declareChar(ad3) -> TAC:
    return TAC("TACDECLARECHAR", "", "", ad3)


def declareInt(ad3) -> TAC:
    return TAC("TACDECLAREINT", "", "", ad3)


def declareString(ad3) -> TAC:
    return TAC("TACDECLARESTRING", "", "", ad3)


def declareArrayInt(ad1, ad3) -> TAC:
    return TAC("TACDECLAREARRAYINT", ad1, "", ad3)


def declareArrayChar(ad1, ad3) -> TAC:
    return TAC("TACDECLAREARRAYCHAR", ad1, "", ad3)


def label(ad3) -> TAC:
    return TAC("TACLABEL", "", "", ad3)


def TAClist2JSON(TAClist: list):
    from json import dumps
    return dumps([x.toDict() for x in TAClist])


def printTAClist(TAClist: list):
    print(TAClist2JSON(TAClist).replace("None", "null"))


class newAddress:
    def __init__(self):
        self.cnt = 0

    def __call__(self, *args, **kwargs):
        self.cnt += 1
        if len(args)==0:
            return "ti" + str(self.cnt)
        if str(args[0]).lower() == "int":
            return "ti" + str(self.cnt)
        elif str(args[0]).lower() == "char":
            return "tc" + str(self.cnt)

class newLabel:
    def __init__(self):
        self.cnt = 0

    def __call__(self, *args, **kwargs):
        self.cnt += 1
        return "label"+str(self.cnt)


getAddress = newAddress()
getLabel = newLabel()

File: compiler/src/TAC.cpp
//
// Created by zhaoyiping on 2020/5/2.
//

#include "TAC.h"
const address addrUNDEFINED = "";
TAC::TAC()
    : op(TACUNDEFINED), ad1(addrUNDEFINED), ad2(addrUNDEFINED), ad3(addrUNDEFINED) {

}
TAC::TAC(TACop op, address ad1, address ad2, address ad3)
    : op(op), ad1(ad1), ad2(ad2), ad3(ad3) {

}
TAC::TAC(TACop op, address ad1, address ad3)
    : op(op), ad1(ad1), ad2(addrUNDEFINED), ad3(ad3) {

}
TAC::TAC(TACop op, address ad3)
    : op(op), ad1(addrUNDEFINED), ad2(addrUNDEFINED), ad3(ad3) {

}
TAC::TAC(TACop op)
    : op(op), ad1(addrUNDEFINED), ad2(addrUNDEFINED), ad3(addrUNDEFINED) {

}
json TAC::toJSON() {
  json ret;
  ret["op"] = toString(op);
  if (ad1 != addrUNDEFINED)ret["ad1"] = ad1;
  else ret["ad1"] = "";
  if (ad2 != addrUNDEFINED)ret["ad2"] = ad2;
  else ret["ad2"] = "";
  if (ad3 != addrUNDEFINED)ret["ad3"] = ad3;
  else ret["ad3"] = "";
  return ret;
}
bool operator==(const TAC &a, const TAC &b) {
  return a.op == b.op && a.ad1 == b.ad1 && a.ad2 == b.ad2 && a.ad3 == b.ad3;
}
AddressGenerator::AddressGenerator() : tmpCnt(0) {
  allAddress.clear();
}
address AddressGenerator::newTmpAddr(Tokentype tokentype) {
  std::string addr = "t";
  if (tokentype == CHAR)addr += "c";
  else addr += "i";
  addr += std::to_string(tmpCnt + 1);
  allAddress.emplace_back(addr);
  tmpCnt++;
  return allAddress.back();
}
LabelGenerator::LabelGenerator() : tmpCnt(0) {
  allLabel.clear();
}
address LabelGenerator::newLabel() {
  allLabel.emplace_back(address("label" + std::to_string(tmpCnt + 1)));
  tmpCnt++;
  return allLabel.back();
}
void LabelGenerator::addLabel(address funcName) {
  allLabel.emplace_back(funcName);
}
TransInfo::TransInfo() :
    labelGenerator(nullptr),
    addressGenerator(nullptr),
    returnLabel(addrUNDEFINED),
    thenLabel(addrUNDEFINED),
    elseLabel(addrUNDEFINED) {}
TransInfo::TransInfo(std::shared_ptr<LabelGenerator> labelGenerator,
                     std::shared_ptr<AddressGenerator> addressGenerator,
                     address returnLabel,
                     address thenLabel,
                     address elseLabel) :
    labelGenerator(labelGenerator),
    addressGenerator(addressGenerator),
    returnLabel(returnLabel),
    thenLabel(thenLabel),
    elseLabel(elseLabel) {}

std::string toString(TACop tacOp) {
  static std::unordered_map<TACop, std::string> ma;
  if (ma.empty()) {
#define TAC2String(x) ma[x]=#x;
    TAC2String(TACADD);
    TAC2String(TACSUB);
    TAC2String(TACMUL);
    TAC2String(TACDIV);
    TAC2String(TACMOV);
    TAC2String(TACGETARR);
    TAC2String(TACSETARR);
    TAC2String(TACJLE);
    TAC2String(TACJGE);
    TAC2String(TACJNE);
    TAC2String(TACJEQ);
    TAC2String(TACJG);
    TAC2String(TACJL);
    TAC2String(TACCALL);
    TAC2String(TACRET);
    TAC2String(TACSETRET);
    TAC2String(TACGETRET);
    TAC2String(TACSETARG);
    TAC2String(TACGETARG);
    TAC2String(TACJ);
    TAC2String(TACREADINT);
    TAC2String(TACREADCHAR);
    TAC2String(TACWRITEINT);
    TAC2String(TACWRITECHAR);
    TAC2String(TACWRITESTRING);
    TAC2String(TACDECLAREINT);
    TAC2String(TACDECLARECHAR);
    TAC2String(TACDECLAREINTARG);
    TAC2String(TACDECLARECHARARG);
    TAC2String(TACDECLARESTRING);
    TAC2String(TACDECLAREARRAYINT);
    TAC2String(TACDECLAREARRAYCHAR);
    TAC2String(TACLABEL);
    TAC2String(TACUNDEFINED);
#undef TAC2String
  }
  return ma[tacOp];
}

TACop toTACop(std::string s) {
  static std::unordered_map<std::string, TACop> ma;
  if (ma.empty()) {
#define String2TAC(x) ma[#x]=x;
    String2TAC(TACADD);
    String2TAC(TACSUB);
    String2TAC(TACMUL);
    String2TAC(TACDIV);
    String2TAC(TACMOV);
    String2TAC(TACGETARR);
    String2TAC(TACSETARR);
    String2TAC(TACJLE);
    String2TAC(TACJGE);
    String2TAC(TACJNE);
    String2TAC(TACJEQ);
    String2TAC(TACJG);
    String2TAC(TACJL);
    String2TAC(TACCALL);
    String2TAC(TACRET);
    String2TAC(TACSETRET);
    String2TAC(TACGETRET);
    String2TAC(TACSETARG);
    String2TAC(TACGETARG);
    String2TAC(TACJ);
    String2TAC(TACREADINT);
    String2TAC(TACREADCHAR);
    String2TAC(TACWRITEINT);
    String2TAC(TACWRITECHAR);
    String2TAC(TACWRITESTRING);
    String2TAC(TACDECLAREINT);
    String2TAC(TACDECLARECHAR);
    String2TAC(TACDECLARESTRING);
    String2TAC(TACDECLAREARRAYINT);
    String2TAC(TACDECLAREARRAYCHAR);
    String2TAC(TACLABEL);
    String2TAC(TACUNDEFINED);
#undef String2TAC
  }
  if (ma.count(s))
    return ma[s];
  return TACUNDEFINED;
}
bool isJump(TAC tac) {
  return tac.op == TACJ ||
      tac.op == TACJL ||
      tac.op == TACJLE ||
      tac.op == TACJG ||
      tac.op == TACJGE ||
      tac.op == TACJNE ||
      tac.op == TACJEQ;
}
bool isWrite(TAC tac) {
  return tac.op == TACWRITEINT || tac.op == TACWRITESTRING || tac.op == TACWRITECHAR;
}
bool isRead(TAC tac){
  return tac.op==TACREADINT||tac.op==TACREADCHAR;
}
bool isFunctionLabel(address label) {
  return label.substr(0, 5) != "label";
}
bool isRead(address addr);
bool isDeclare(TAC tac) {
  switch (tac.op) {
    case TACDECLAREINT:
    case TACDECLARECHAR:
    case TACDECLAREINTARG:
    case TACDECLARECHARARG:
    case TACDECLAREARRAYINT:
    case TACDECLAREARRAYCHAR:
    case TACDECLARESTRING:return true;
    default:return false;
  }
}

bool isAddress(address addr) {
  return addr[0] == 'l' || addr[0] == 'g' || addr[0] == 't';
}

bool isGlobal(address addr) {
  return addr[0] == 'g';
}

bool isTmp(address addr) {
  return addr[0] == 't';
}

bool isLocal(address addr) {
  return addr[0] == 'l';
}

bool isString(address addr) {
  return addr[0] == '"' && addr.back() == '"';
}

bool isInt(address addr) {
  return addr[1] == 'i';
}

bool isChar(address addr) {
  return addr[1] == 'c';
}
